File: DSA/Stack/stack.cpp
// stack implementation in cpp for generic types as well
//
#include <iostream>
#define STACK_SIZE 100
template <class T>

class Stack {
private:
  int top;
  T *nodes;

public:
  Stack();
  int empty(void);
  T pop(void);
  T peek(void);
  void push(const T &);
  ~Stack();
};
template <class T> Stack<T>::~Stack() { delete nodes; }
template <class T> Stack<T>::Stack() {
  top = -1;
  nodes = new T[STACK_SIZE];
}

template <class T> void Stack<T>::push(const T &j) {
  if (top == STACK_SIZE - 1) {
    std::cout << "Stack overflow\n";
  }
  nodes[++top] = j;
}

template <class T> T Stack<T>::pop() {
  T p;
  if (empty()) {
    std::cout << "Stack underflow\n";
    return p;
  }
  return nodes[top--];
}

template <class T> int Stack<T>::empty() { return top == -1; }
template <class T> T Stack<T>::peek() {
  T p{};
  if (!empty()) {
    return nodes[top];
  }
  std::cout << "Stack underflow\n";
  return p;
}

int main() {
  Stack<char> sc;
  sc.push('a');
  std::cout << sc.peek() << std::endl;
  sc.pop();
  std::cout << sc.peek() << std::endl;
  Stack<int> si;
  si.push(1);
  std::cout << si.peek() << std::endl;
}

File: DSA/Stack/stack.rs
struct Stack<T> {
    stack: Vec<T>,
}
impl<T> Stack<T> {
    fn new() -> Self {
        Stack { stack: Vec::new() }
    }
    fn pop(&mut self) -> Option<T> {
        self.stack.pop()
    }
    fn push(&mut self, item: T) {
        self.stack.push(item);
    }
    fn empty(&self) -> bool {
        self.stack.is_empty()
    }
    fn peek(&self) -> Option<&T> {
        self.stack.last()
    }
}
fn main() {
    let mut stack: Stack<i32> = Stack::new();
    stack.push(2);
    println!("{}", stack.peek().unwrap());
}
